### PR TITLE
[TIMOB-17145]Add support to HTTPClient for PATCH requests

### PR DESCRIPTION
**Pull request by kabomi**
Ticket url: https://jira.appcelerator.org/browse/TIMOB-17145
#### Screenshot showing the problem

patchmethodnotsupported
#### Screenshot showing the expected outcome

patchmethodsupported
#### index.js

``` javascript
function doClick(e) {
    doPatch();
}

function doPatch(data){
    var url = "https://api.github.com/notifications/threads/65269820";
    var xhr = Ti.Network.createHTTPClient();
    console.log('BEFORE PATCH');
    xhr.open('PATCH', url);

    xhr.onload = function(e){
        console.log('[onload] success');
    };
    xhr.onerror =function(e){
        console.log('[error] success');
    };
    xhr.onreadystatechange = function(e){
        switch(this.readyState) {
             case 0:
             // after HTTPClient declared, prior to open()
             // though Ti won't actually report on this readyState
                            Ti.API.info('case 0, readyState = '+this.readyState);
             break;
             case 1:
             // open() has been called, now is the time to set headers
                            Ti.API.info('case 1, readyState = '+this.readyState);
             break;
             case 2:
             // headers received, xhr.status should be available now
                            Ti.API.info('case 2, readyState = '+this.readyState);
             break;
             case 3:
             // data is being received, onsendstream/ondatastream being called now
                            Ti.API.info('case 3, readyState = '+this.readyState);
                            console.log(this);
             break;
             case 4:
             // done, onload or onerror should be called now
                            Ti.API.info('case 4, readyState = '+this.readyState);
             break;
        };
    };
    xhr.setRequestHeader('Authorization','token ' + '1e95427bc6e6932d37e8f9233ed5cd01dd52ed00');
    xhr.send();
    console.log('SENT');
}

$.index.open();

```

### CODE REVIEW

**Comment by ajwhite** [comment]
+1 really could use this

File: android/modules/network/src/java/ti/modules/titanium/network/TiHTTPClient.java [hunk 1]
```diff
@@ -1171,7 +1170,7 @@ public void send(Object userData) throws MethodNotSupportedException
 				boolean queryStringAltered = false;
 				for (String key : data.keySet()) {
 					Object value = data.get(key);
-					if (isPostOrPut && (value != null)) {
+					if (isPostOrPutOrPatch && (value != null)) {
```
Comment (ajwhite) [inline android/modules/network/src/java/ti/modules/titanium/network/TiHTTPClient.java]: I personally hate this, technically any request is permitted a request body with the HTTP spec. I've seen API's that support `DELETE` with request body parameters.

I suppose this is not your beef though, and should be addressed separately.

Looks good :+1:

**Comment by ajwhite** [comment]
Pssst -- make sure to sign the CLA https://cla.appcelerator.com/

**Comment by kabomi** [comment]
Ok, I signed again through github. I already had one signed by Neeraj Gupta though. If there is anything more that I can do please let me know :-)

**Comment by ajwhite** [comment]
Strange! I think this looks good. I'm not a maintainer of the project, just a member of the community as well eager for this support to be added.

I'm Atticus White in the comments on the ticket you're addressing https://jira.appcelerator.org/browse/TIMOB-17145

**Comment by ashcoding** [comment]
@kabomi Hi! Thanks for your pull request. I'm currently looking and reviewing it. There's only one line that I feel that needs changes. Once that's done, everything should be good to go!

File: android/modules/network/src/java/ti/modules/titanium/network/TiDefaultHttpRequestFactory.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+		super();
+	}
+
+	@Override
+	public HttpRequest newHttpRequest(final RequestLine requestline)
+			throws MethodNotSupportedException {
+
+		String method = requestline.getMethod();
+
+		if (requestline == null) {
```
Comment (ashcoding) [inline android/modules/network/src/java/ti/modules/titanium/network/TiDefaultHttpRequestFactory.java]: Could you put the null checking before the getMethod()?

Comment (kabomi) [inline android/modules/network/src/java/ti/modules/titanium/network/TiDefaultHttpRequestFactory.java]: Yeah, thank you

**Comment by ashcoding** [comment]
Code reviewed and functionally tested. **PR Accepted**
:+1:  @kabomi Thank you.

**Pull request merged**